bpo-43972: Set content-length to 0 for http.server.SimpleHTTPRequestHandler 301s

sirosen opened the pull request:
When `http.server.SimpleHTTPRequestHandler` sends a 301 (Moved Permanently) due to a missing file, add a Content-Length of 0. This improves the behavior for certain clients.

<!-- issue-number: [bpo-43972](https://bugs.python.org/issue43972) -->
https://bugs.python.org/issue43972
<!-- /issue-number -->

sirosen committed:
Set content-length for simple http server 301s

When http.server.SimpleHTTPRequestHandler sends a 301 (Moved
Permanently) due to a missing file, it does not set a Content-Length
of 0. Unfortunately, certain clients can be left waiting for the
connection to be closed in this circumstance, even though no body
will be sent. At time of writing, both curl and Firefox demonstrate
this behavior.

orsenthil commented on Lib/http/server.py:
```diff
@@ -689,6 +689,7 @@ def send_head(self):
                              parts[3], parts[4])
                 new_url = urllib.parse.urlunsplit(new_parts)
                 self.send_header("Location", new_url)
+                self.send_header("Content-Length", "0")
```
Please add a test case for this change.

sirosen replied:
I've added a single check to one of the existing http server tests.

If something else is warranted in this case, I'm more than happy to do it, but I'd need some direction.
The added check seems sufficient to me and in-keeping with the way those existing tests are written.

sirosen committed:
Test Content-Length on simple http server redirect

When serving a redirect, the SimpleHTTPRequestHandler will now send
`Content-Length: 0`. Several tests for http.server already cover
various behaviors and checks including redirection. This change only
adds one check for the expected Content-Length on the simplest case
for a redirect.

sirosen committed:
Add news entry for SimpleHTTPRequestHandler fix

orsenthil committed:
Clarify the specific kind of 301

orsenthil committed:
Update 2021-04-30-16-58-24.bpo-43972.Y2r9lg.rst

orsenthil committed:
Update 2021-04-30-16-58-24.bpo-43972.Y2r9lg.rst

orsenthil reviewed (approved):
LGTM.

orsenthil reviewed (approved):
LGTM.
